### ISSUE TITLE
Batch inference size

### ISSUE THREAD

**Issue by zjjzeng** [ultralytics/ultralytics#22898]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi Ultralytics team

I am trying to do a YOLO11 batch inference for my varied size source input. They are 512*512 `imgsz` and just 1 channel numpy matrices. For most cases, the code `results = model.predict(source, imgsz=512, half=True)` works well, but for the large input like a list of around 4000, the program will be killed by OOM, and my OS (ubuntu) also be stuck. My device is Nvidia DGX Spark, I believe the GPU memory is plenty. So I wonder the exact memory malloc strategy of the batch inference. And if I need maximum speed for my program, is it worthy using as large batch size as possible?

Look forward to your reply. Thanks in advance.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zjjzeng, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly — an Ultralytics engineer will also review your issue and follow up soon 🙂

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since your question involves batch inference behavior and potential memory issues, it would be very helpful if you could clarify whether this is a suspected 🐛 Bug Report or more of a usage/performance ❓ Question.

-If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) that shows:
  - The exact code snippet you are running for batch inference
  - How `source` is constructed (e.g., list/array shape, dtypes, device)
  - Your environment details (OS, Python, PyTorch, CUDA, GPU model, `ultralytics` version)
  - Any logs, tracebacks, or system messages before the OOM kill  
  This will help us reliably reproduce and investigate the behavior you’re seeing.

-If this is a custom training or inference ❓ Question, please provide as much detail as possible, including:
  - A small sample of your input data (shape, channels, type; for example how your 1‑channel numpy matrices are passed to the model)
  - The full `model.predict(...)` call (all arguments you set)
  - Any profiling or memory observations you’ve made  
  Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), as many of the same principles apply to efficient inference.

You can also join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community

## Upgrade

Before we dig deeper, please ensure you have upgraded to the latest `ultralytics` package (and dependencies) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/), and then re-test your batch inference to confirm whether the issue persists:

```bash
pip install -U ultralytics
```

You can also check the [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) to verify your environment is aligned.

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Running your MRE in one of these verified environments can help us quickly determine whether the behavior is environment-specific.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share an MRE and the extra details above, it will greatly help the engineering team provide a precise follow-up on your batch inference and memory behavior question 💡

**Comment by Y-T-G** [comment]
Are you getting OOM from system memory or CUDA?

There's no benefit in increasing batch size after a certain point because the GPU saturates. So a batch of 4000 is unnecessary and overkill.

**Comment by zjjzeng** [comment]
I found `Out of memory: Kill process` from dmesg, the program only showed SIKILL exit with -9 in the console. But I am sure the issue raised at the batch inference line, even though my device has 128G memory/GPU memory.

For the batch size, any guidance on the certain point, what is the best practice for some large input source?

**Comment by Y-T-G** [comment]
You can check by doubling the batch and testing the inference time. Starting from 16. Once inference time per image stops improving, you know the GPU is saturated.
